lib/segment/tests/integration/multivector_quantization_test.rs

```diff
@@ -0,0 +1,361 @@
+};
+use segment::entry::entry_point::SegmentEntry;
+use segment::fixtures::payload_fixtures::{random_int_payload, random_multi_vector};
+use segment::index::hnsw_index::graph_links::GraphLinksRam;
+use segment::index::hnsw_index::hnsw::HNSWIndex;
+use segment::index::hnsw_index::num_rayon_threads;
+use segment::index::{PayloadIndex, VectorIndex};
+use segment::json_path::path;
+use segment::segment_constructor::build_segment;
+use segment::types::{
+    BinaryQuantizationConfig, CompressionRatio, Condition, Distance, FieldCondition, Filter,
+    HnswConfig, Indexes, MultiVectorConfig, Payload, PayloadSchemaType, ProductQuantizationConfig,
+    QuantizationSearchParams, Range, ScalarQuantizationConfig, SearchParams, SegmentConfig,
+    SeqNumberType, VectorDataConfig, VectorStorageType,
+};
+use segment::vector_storage::quantized::quantized_vectors::QuantizedVectors;
+use segment::vector_storage::query::{ContextPair, DiscoveryQuery, RecoQuery};
+use serde_json::json;
+use tempfile::Builder;
+
+const MAX_EXAMPLE_PAIRS: usize = 4;
+const MAX_VECTORS_COUNT: usize = 3;
+
+enum QueryVariant {
+    Nearest,
+    RecommendBestScore,
+    Discovery,
+}
+
+enum QuantizationVariant {
+    Scalar,
+    PQ,
+    Binary,
+}
+
+fn random_vector<R: Rng + ?Sized>(rnd: &mut R, dim: usize) -> MultiDenseVector {
+    let count = rnd.gen_range(1..=MAX_VECTORS_COUNT);
+    let mut vector = random_multi_vector(rnd, dim, count);
+    // for BQ change range to [-0.5; 0.5]
+    vector.flattened_vectors.iter_mut().for_each(|x| *x -= 0.5);
+    vector
+}
+
+fn random_discovery_query<R: Rng + ?Sized>(rnd: &mut R, dim: usize) -> QueryVector {
+    let num_pairs: usize = rnd.gen_range(1..MAX_EXAMPLE_PAIRS);
+
+    let target = random_vector(rnd, dim).into();
+
+    let pairs = (0..num_pairs)
+        .map(|_| {
+            let positive = random_vector(rnd, dim).into();
+            let negative = random_vector(rnd, dim).into();
+            ContextPair { positive, negative }
+        })
+        .collect_vec();
+
+    DiscoveryQuery::new(target, pairs).into()
+}
+
+fn random_reco_query<R: Rng + ?Sized>(rnd: &mut R, dim: usize) -> QueryVector {
+    let num_examples: usize = rnd.gen_range(1..MAX_EXAMPLE_PAIRS);
+
+    let positive = (0..num_examples)
+        .map(|_| random_vector(rnd, dim).into())
+        .collect_vec();
+    let negative = (0..num_examples)
+        .map(|_| random_vector(rnd, dim).into())
+        .collect_vec();
+
+    RecoQuery::new(positive, negative).into()
+}
+
+fn random_query<R: Rng + ?Sized>(variant: &QueryVariant, rnd: &mut R, dim: usize) -> QueryVector {
+    match variant {
+        QueryVariant::Nearest => random_vector(rnd, dim).into(),
+        QueryVariant::Discovery => random_discovery_query(rnd, dim),
+        QueryVariant::RecommendBestScore => random_reco_query(rnd, dim),
+    }
+}
+
+fn sames_count(a: &[Vec<ScoredPointOffset>], b: &[Vec<ScoredPointOffset>]) -> usize {
+    a[0].iter()
```

Comment:
We ask for the top 5 in the test but then just compare the first result between plain and indexed.

I find this implementation a bit misleading given the name.
Where the results really bad for the rest of the tops?

Reply:
Search with exact flag uses original vectors instead of quantized. That's why I compare hnsw search with plain - it's a comparison between between quantized and non-quantized data